Implement Vim's tag stack

zenazn opened the pull request:
cc @ConradIrwin @AidanV (thank you for pairing with me!)

Gosh, this turned out to be harder than I thought! But I'm really happy with how things turned out—it's a much better solution than what I had on Tuesday.

I'd recommend reading this commit-by-commit. I tried to leave nice commit messages to make it easier to follow along.

At a high level:
- This ~exactly implements Vim's tag stack behavior, at least as I understand and use it. We're missing (1) a UI, and (2) some stuff like navigating between ambiguous choices, which Zed has a different solve for anyways
- I've integrated it tightly with Zed's navigation history (forwards/back). You can think of navigating to older and newer tags like "super-back" and "super-forward"
- Nothing about this implementation is Vim specific. Anyone can use it!
- Except for the fact that splits now fork navigation (and tag) history, and the exact details of which navigations result in history items, I am not aware of behavior differences for folks who don't use the feature.

One thing I haven't looked into: Making `<C-t>` in Vim use `pane: Go to older tag` instead of `pane: Go back`. 

Closes #14206

Release Notes:
- ??? I don't know what to write here! Suggestions welcome (or Just Fix It™)

zenazn committed:
pane: clone for NavHistory

This commit makes it clear how much the changes here spill out into the
rest of the repo. Being able to clone NavHistory allows us to fork
history when splitting (implementation in next commit)

zenazn committed:
pane: fork navigation history on split

This change is hopefully mildly useful for everyone (you can now use
splits to "explore" history without risking giving up on your current
state), and will be very helpful in a few moments for implementing the
Vim tag stack

zenazn committed:
pane: implement navigating tag history

Nothing is adding entries to the tag stack yet, but were they to exist
this is how we'd navigate between entries:
- A good mental model: the tag stack is "super-back" and "super-forward"
  in navigation history
- Every time we go-to-definition, that navigation (i.e., the edge
  between those two history states) gets pushed onto the stack [not yet
  implemented; see next commit]
- When we go back to an older tag, we navigate back in history (as if
  we'd hit "back" a lot) to the moment *before* we navigated to the
  most recent definition
- When we go forward to a newer tag, we navigate forward in history to
  the moment *after* we navigated to the next definition
- In the case that the tag stack refers to history states that are no
  longer part of navigation history (e.g., they have aged off the back
  of the stack, or have been truncated off a forward stack because the
  user has started exploring a different branch of history), we navigate
  to the expected positions anyways, pretending as if they were still
  present in navigation history

After much thought, I think this is the best way to work Vim's tag stack
into Zed's concept model. I think it's especially nice that:
- There's nothing Vim-specific about anything in the implementation,
  even though it's inspired by (and has more-or-less the same semantics
  as) Vim
- There are virtually no changes for users that don't want to interact
  with the tag stack (two minor changes: the exact history entries that
  get created are different, and navigation history forks on pane split)
- It's possible (and nice, even!) to bounce back and forth between
  back/forward and older/newer tags. They bring different things to the
  table, and work very well together.

The hardest part of the implementation is stably addressing moments in
navigation history:
- The current view state is part of navigation history (especially if
  e.g., you've hit "back" several times and there are navigation
  elements both before and after the current moment in history), but is
  not reified anywhere
- There are no stable identifiers anywhere. Even `(pane, location)` can
  wobble around a little, because a small move to the cursor won't
  create a navigation history entry, but *will* impact what position is
  put (back) onto navigation history once we scroll past it

The solution I came up with is to number navigation history items
starting from 0, and rely on the fact that the common way to manipulate
navigation history is adding and removing elements at the end. This has
one major caveat, which is that there *is* a circumstance in which we
manipulate "middle" entries in history: when we e.g., delete a file. In
this case, instead of removing those entries from history, we leave a
tombstone. This eliminates the need for re-numbering, which makes the
implementation easier.

zenazn committed:
editor: actually-always push to nav history

Previously, even when "always" was set, a nav history entry was not
added if (e.g.,) the definition was on the same line as the place we
were jumping from. But always means always! In particular, this makes it
much easier to interact with navigation history from code, because its
behavior is more predictable.

Also, abstract out `navigation_data`, because we'll need it in just a
sec.

zenazn committed:
editor: push tags when jumping to hover link

Whenever we're navigating to a hover link (which is how features like
go-to-definition are implemented), add that to the tag stack. This is
the equivalent of jumping to a tag in Vim.

We could hook in one layer "up," in `go_to_definition_of_kind`, but I
chose to hook into hover links because:
- Command-click is basically the same as go-to-definition, and it's nice
  that the two are treated identically. Tag stack users can bounce
  seamlessly between navigating with their mouse and keyboard.
- On a practical level, all the information I wanted was in hover links,
  and would have been a pain to pipe around to where I needed it

This commit additionally calculates and stores a "tag," which is the
identifier we jumped to the definition of. This tag isn't used yet, but
is helpful when debugging, and would form the basis of a future UI
displaying the tag stack.

zenazn commented on crates/editor/src/editor.rs:
```diff
@@ -16493,9 +16532,12 @@ impl Editor {
-                    buffer: target_buffer_handle,
-                    range,
+                LocationLink {
+                    origin: None,
```
I don't know enough about this code to know if we can get an `origin` for this jump

zenazn commented on crates/editor/src/editor.rs:
```diff
@@ -16380,21 +16397,31 @@ impl Editor {
+                    ) else {
+                        return Navigated::No;
+                    };
+                    ed.read(cx).create_tag_stack_entry(tag, cx);
```
We chatted on Tuesday about what to do in the case where go-to-definition is ambiguous, and Zed opens a multibuffer of all of the options. I decided that a Good Enough™ solution was to treat the multibuffer itself as the jump target. We can improve this if we'd like, but honestly it feels fine to me

AidanV committed:
Fix missing nav hist to unopened files + origin name fix

AidanV commented:
Great work! I have a crude list of fast follows for commands and keybinds (some of which you have already mentioned) that I think we should do. I would love to knock some of these out after this gets merged.
- [ ] Go to definition should match vim behavior of find nearest keyword to the right if not on a keyword
- [ ] ctrl-] 
- [ ] g]
- [ ] {count}Ctrl-T and :[count]po[p][!]
- [ ] :[count]ta[g][!]
- [ ] :tags
- [ ] {Visual}Ctrl-]
- [ ] :ts[elect][!] [name]
- [ ] :sts[elect][!] [name]
- [ ] {Visual}g]
- [ ] :tj[ump][!] [name]
- [ ] :stj[ump][!] [name]
- [ ] g CTRL-]
- [ ] {Visual}g CTRL-]
- [ ] :[count]tn[ext][!]
- [ ] ]t
- [ ] :[count]tp[revious][!]  /  :[count]tN[ext][!]  /  [t
- [ ] :[count]tr[ewind][!]  /  :[count]tf[irst][!]  /  [T
- [ ] :tl[ast][!]  /  ]T
- [ ] :lt[ag][!] [name]

This mostly came from https://vimdoc.sourceforge.net/htmldoc/tagsrch.html

zenazn commented:
Sorry for the delay—was on vacation this past week.

I rebased past #38707 (which fixes an issue I've seen a lot while testing go-to-definition—thank you!) and added @AidanV's patch. What else would be helpful to move this change forward?

ConradIrwin commented:
Thanks! I'll try and take more of a look in the next few days (though I am quite behind on code review, so I'm going to be a bit slower than I want to be)

zenazn committed:
pane: Send + Sync for clippy

zenazn commented:
@ConradIrwin to follow up on this, what's a good next step? I could:
- Schedule another office hours with you to discuss
- Write a new PR that narrowly implements the behavior I most want (go-back-to-older-tag), which I believe would be less complex
- Try to take the PR in a different direction that you prefer
- Close this PR

I know you're busy and that you want to make sure you understand the code that you'll have to maintain going forward, but from my perspective I'd like to reach some finality to my efforts here

ConradIrwin commented:
Thanks again for your patience.

I finally got time to sit down with this, and the result is https://github.com/zed-industries/zed/commit/b69a2ea200561de9de0edf4d0c4428ee887fb918.

I realized that in vim, the tag stack and the jump list are completely independent (i.e. going back in the tag stack pushes to the jump list), and so trying to merge the two stacks was a mistake (sorry about that, I think it was my idea).

There's still work to do (either on this branch or yours), but I think the direction on the tag-stack-2 branch is less intertwined and a bit easier to follow (though I don't really love the `start/finish` jumps not being close together).

If you want to take it over, that would be great; otherwise I'll continue pushing on this when I get more time.

Todo:
* [ ] Refactor the navigate_tag_stack to use the NavigationMode? Not sure I want this, but don't love the duplication in workspace for navigate_tag_history
* [ ] Clean up names
* [ ] Add tests
* [ ] Add GoToNewerTag (so we can use `:pop` and `:tag`, though actually implementing the commands can be a separate PR).
* [ ] It seems like GoToDefinition falls back to FindAllReferences if you do it from the definition itself; in this case we don't push to the tag stack, should we?

ConradIrwin commented:
Closing this for now, sorry I didn't get to finishing up my implementation. If someone wants to pick that up, it would be fabulous.
